### ISSUE TITLE
Create FileSetting

### ISSUE THREAD

**Issue by rastytheamateur** [pawamoy/django-appsettings#77]
Create a setting, which would take a file path. It's validation would be to open a file. Prototype:

```
class FileSetting(StringSetting):
    """String with file path setting."""

    def validate(self, value):
        """Check if file exists and is readable."""
        open(value)
```

stinovlas suggested to add a permission check.

The value could return either just a string path or a file descriptor (file object) from `open()` .

**Comment by ziima** [comment]
The check for `read` permission may not be sufficient to check the file permissions. Also setting may point to a file which doesn't yet exist, which can be also a valid case, e.g. a cache directory.

As a default I would only return the file path, since the usage of the file may vary a lot. Returning a `Path` object might be even better.

**Comment by pawamoy** [comment]
Returning a file descriptor seems a bit fragile (when should it be closed? will the setting still work after closing the fd? what if the file is truncated / moved?).

I was thinking to return a `Path` object as well instead. With a `Path` you can do nice things like

```python
with settings.cache_file.open("w"):
    # stuff
```
```python
settings.cache_dir.mkdir(parents=True, exist_ok=True)
cache_file = settings.cache_dir / "00001"
# etc
```

**Comment by rastytheamateur** [comment]
So the only special thing is to return a `Path` object from `transform` after all? Should we at least add an `exists` attribute into the constructor and check if the file really exists when the attribute value is `True`? Or moreover an attribute (and validation check) for permissions?

**Comment by stinovlas** [comment]
I still believe, that check for existence and also check for specific permissions is useful. I.e. if I want to set an input or configuration file then I want that file to exist and I also want to have read permissions. Or, I could set a log file and I'd like to have write permissions for that path.

**Comment by pawamoy** [comment]
But what if the asked permission are not set? What if you need write permission but you don't have it? Will you `chmod` the file somehow? Shouldn't you directly try instead to read/write the file and catch a `PermissionError` to log something ("ask forgiveness not permission")? Or do you simply want to fail and warn the user as early as possible?

**Comment by stinovlas** [comment]
It's true that the existence of the file or even its permissions can change after the configuration is loaded, so the *fail early* policy still doesn't guarantee anything. I see a benefit in protecting against misspelling filenames and such. Of course, we can only check permissions on existing files.

**Comment by rastytheamateur** [comment]
I [tried to do something](https://github.com/rastytheamateur/django-appsettings/compare/master...create-file-setting), but I am still not sure if it satisfies all the mentioned scenarios. Can you, please, check the diff and tell me if there is something obviously wrong before I create a pull request? Or would you add something?

Now I have ``exists=True, read_perm=True, write_perm=False`` attributes, which indicate whether or not should we validate individual cases. In the docstring there is an information that it is not guaranteed nothing changes between the validation and the usage...

**Comment by ziima** [comment]
Since we have validators, I'd move the checks into them and wouldn't keep them in the `FileSettings` itself.

**Comment by stinovlas** [comment]
Also, it might be better to use [standard open mode syntax](https://docs.python.org/3/library/functions.html#open) rather than creating `read_perm`, `write_perm` and potentially other validators. You can just create `check_perm` validator with attribute `mode`, which will be in the same format that `open` uses.

**Comment by ziima** [comment]
Fixed by #81
